modules/op-results-search-records.adoc

```diff
@@ -0,0 +1,73 @@
+// This module is included in the following assembly:
+//
+// *cicd/pipelines/using-tekton-results-for-openshift-pipelines-observability.adoc
+
+:_content-type: PROCEDURE
+[id="results-search-records_{context}"]
+= Searching for records
+
+[role="_abstract"]
+You can search for records using Common Expression Language (CEL) queries. As each record contains full YAML information for a pipeline run or task run, you can find records by many different criteria.
+
+.Prerequisites
+
+* You installed {tekton-results}.
+* You installed the `opc` utility and prepared its environment to quety {tekton-results}.
+
+.Procedure
+
+* To search for records using a CEL query, run the following command:
++
+[source, terminal]
+----
+$ opc records list --addr ${RESULTS_API} --filter=<CEL query> <namespace-name>/result/-
+----
++
+Replace `<namespace-name>` with the namespace in which the pipeline runs or task runs were created. Alretnatively, to search the records within a single result, use the following command:
++
+[source, terminal]
+----
+$ opc records list --addr ${RESULTS_API} --filter=<CEL query> <result-name>
+----
++
+Replace `<result-name>` with the full name of the result.
+
+
+.Example CEL queries for records
+|===
+| Purpose | CEL query
+
+| Records of all task runs or pipeline runs that failed
+| `!(data.status.conditions[0].status == 'True')`
+
+| Records where the name of the `TaskRun` or `PipelineRun` custom resource (CR) was `run1`
+| `data.metadata.name == 'run1'`
+
+| Records for all task runs that were started by the `PipelineRun` CR named `run1`
+| `data.metadata.labels['tekton.dev/pipelineRun'] == 'run1'`
```

Comment:
@avinal are you sure that this is supposed to work? When I run `opc results records list ppitonak/results/d370d70f-e4f2-465f-8bb8-20883ef4bb90 -o json`, I get this (truncated)
```
{
  "records":  [
    {
      "name":  "ppitonak/results/d370d70f-e4f2-465f-8bb8-20883ef4bb90/records/3d2e93f5-0982-3c6f-8eb3-b935e2bdd0f9",
      "id":  "68cd1b6b-0c0d-4e47-801e-d96ab7b1435f",
      "uid":  "68cd1b6b-0c0d-4e47-801e-d96ab7b1435f",
      "data":  {
        "type":  "results.tekton.dev/v1alpha2.Log",
        "value":  "eyJraW5kIjogIkxvZyIsICJzcGVjIjogeyJ0eXBlIjogIkZpbGUiLCAicmVzb3VyY2UiOiB7InVpZCI6ICI3NDY5MjY5OS1iMjA3LTQwMWYtOGQzYy0yZWM1NDdlZWU1NWQiLCAia2luZCI6ICJUYXNrUnVuIiwgIm5hbWUiOiAidW5pbnN0YWxsLWNsdXN0ZXItcnVuLWY4NHdxLXVuaW5zdGFsbCIsICJuYW1lc3BhY2UiOiAicHBpdG9uYWsifX0sICJzdGF0dXMiOiB7InBhdGgiOiAicHBpdG9uYWsvM2QyZTkzZjUtMDk4Mi0zYzZmLThlYjMtYjkzNWUyYmRkMGY5L3VuaW5zdGFsbC1jbHVzdGVyLXJ1bi1mODR3cS11bmluc3RhbGwtbG9nIiwgInNpemUiOiA0MzAxMn0sICJtZXRhZGF0YSI6IHsidWlkIjogIjNkMmU5M2Y1LTA5ODItM2M2Zi04ZWIzLWI5MzVlMmJkZDBmOSIsICJuYW1lIjogInVuaW5zdGFsbC1jbHVzdGVyLXJ1bi1mODR3cS11bmluc3RhbGwtbG9nIiwgIm5hbWVzcGFjZSI6ICJwcGl0b25hayIsICJjcmVhdGlvblRpbWVzdGFtcCI6IG51bGx9LCAiYXBpVmVyc2lvbiI6ICJyZXN1bHRzLnRla3Rvbi5kZXYvdjFhbHBoYTIifQ=="
      },
      "etag":  "68cd1b6b-0c0d-4e47-801e-d96ab7b1435f-1695134714554514038",
      "createdTime":  "2023-09-19T14:45:14.412729Z",
      "createTime":  "2023-09-19T14:45:14.412729Z",
      "updatedTime":  "2023-09-19T14:45:14.554514Z",
      "updateTime":  "2023-09-19T14:45:14.554514Z"
    } ...
```

There is no `.data.metadata.labels` or `.data.metadata.name` etc

Reply:
`data` in filter refers to the base64 decoded `data.value` field in the response. So if you decode, you will get this:
```json
{
  "kind": "Log",
  "spec": {
    "type": "File",
    "resource": {
      "uid": "74692699-b207-401f-8d3c-2ec547eee55d",
      "kind": "TaskRun",
      "name": "uninstall-cluster-run-f84wq-uninstall",
      "namespace": "ppitonak"
    }
  },
  "status": {
    "path": "ppitonak/3d2e93f5-0982-3c6f-8eb3-b935e2bdd0f9/uninstall-cluster-run-f84wq-uninstall-log",
    "size": 43012
  },
  "metadata": {
    "uid": "3d2e93f5-0982-3c6f-8eb3-b935e2bdd0f9",
    "name": "uninstall-cluster-run-f84wq-uninstall-log",
    "namespace": "ppitonak",
    "creationTimestamp": null
  },
  "apiVersion": "results.tekton.dev/v1alpha2"
}
```
You are still partially right that this filter won't work. But that is because you have put the filter on a record of type `Log`. Internally `Log` is too a type of `Record` similar to `PipelineRun and `TaskRun`. So in this case, your query would be wrong.

However, if you run the same filter on records of type `PipeLineRun` and `TaskRun` you get something like this. 
```json
{
  "kind": "PipelineRun",
  "metadata": {
    "uid": "1638b693-844d-4f13-b767-d7d84ac4ab3d",
    "name": "hello",
    "labels": {
      "tekton.dev/pipeline": "hello"
    },
    "namespace": "default",
    "generation": 1,
    "annotations": {
      "results.tekton.dev/record": "default/results/1638b693-844d-4f13-b767-d7d84ac4ab3d/records/1638b693-844d-4f13-b767-d7d84ac4ab3d",
      "results.tekton.dev/result": "default/results/1638b693-844d-4f13-b767-d7d84ac4ab3d",
      "results.tekton.dev/resultAnnotations": "{\"repo\": \"tektoncd/results\", \"commit\": \"1a6b908\"}",
      "results.tekton.dev/recordSummaryAnnotations": "{\"foo\": \"bar\"}",
      "kubectl.kubernetes.io/last-applied-configuration": "{\"apiVersion\":\"tekton.dev/v1beta1\",\"kind\":\"PipelineRun\",\"metadata\":{\"annotations\":{\"results.tekton.dev/recordSummaryAnnotations\":\"{\\\"foo\\\": \\\"bar\\\"}\",\"results.tekton.dev/resultAnnotations\":\"{\\\"repo\\\": \\\"tektoncd/results\\\", \\\"commit\\\": \\\"1a6b908\\\"}\"},\"name\":\"hello\",\"namespace\":\"default\"},\"spec\":{\"pipelineSpec\":{\"tasks\":[{\"name\":\"hello\",\"taskSpec\":{\"steps\":[{\"image\":\"ubuntu\",\"name\":\"hello\",\"script\":\"echo hello world!\"}]}}]}}}\n"
    },
    "resourceVersion": "1567",
    "creationTimestamp": "2023-08-22T09:08:59Z"
  },
  "apiVersion": "tekton.dev/v1beta1"
}
```
I do not have an example for TaskRun, here you can see `metadata.labels["tekton.dev/pipeline"]`. 

In short, if the fields mentioned in filters are not present, you won't find them in the response at all. There must be more records in the list you got, please try filters for a Taskrun/PipelineRun record.

Reply:
OK, now it makes sense but it seem that the filter needs to be properly quoted, otherwise it fails. This is quite user unfriendly so we should at least make it clear in docs.

```
$ opc results records list --filter="data.metadata.name=='acceptance-tests-112-on-414-arm-c8bqj-clone-plumbing-git'" ppitonak/results/16fb205e-092a-45d3-abe8-f2fbc9ea6d1d
Name                                                                                                Type                                    Start                                   Update
ppitonak/results/16fb205e-092a-45d3-abe8-f2fbc9ea6d1d/records/b3e3a8d4-79b6-46e4-8c74-f435966d6e2e  tekton.dev/v1beta1.TaskRun              2023-09-20 18:41:48 +0200 CEST          2023-09-22 08:21:26 +0200 CEST


$ opc results records list --filter=data.metadata.name=='acceptance-tests-112-on-414-arm-c8bqj-clone-plumbing-git' ppitonak/results/16fb205e-092a-45d3-abe8-f2fbc9ea6d1d
ListRecords: rpc error: code = InvalidArgument desc = error parsing filter: ERROR: <input>:1:21: undeclared reference to 'acceptance' (in container '')
 | data.metadata.name==acceptance-tests-112-on-414-arm-c8bqj-clone-plumbing-git
 | ....................^

```

Reply:
Yes, you need to properly quote the filters.